core/search/prefix_filter.cpp

```diff
@@ -25,13 +25,44 @@
 #include <boost/functional/hash.hpp>
 
 #include "shared.hpp"
+#include "filter_visitor.hpp"
 #include "multiterm_query.hpp"
 #include "analysis/token_attributes.hpp"
 #include "index/index_reader.hpp"
 #include "index/iterators.hpp"
 
 NS_ROOT
 
+/*static*/ void by_prefix::visit(
```

Comment:
let's extract template version of 
```
template<typename Visitor>
void visit(const term_reader& reader,
               const bytes_ref& prefix,
               Visitor& visitor) {
// find term
  auto terms = reader.iterator();

  // seek to prefix
  if (IRS_UNLIKELY(!terms) || SeekResult::END == terms->seek_ge(prefix)) {
    return;
  }

  const auto& value = terms->value();
  if (starts_with(value, prefix)) {
    terms->read();

    visitor.prepare(terms);

    do {
      visitor.visit();

      if (!terms->next()) {
        break;
      }

      terms->read();
    } while (starts_with(value, prefix));
  }
}
```
to ensure there are no virtual calls within `phrase_filter.cpp`

then both `by_prefix::visit` and `by_prefix::prepare` will just delegate to template `visit`